[FC-0049] feat: Serialize tag data in OLX for blocks

yusuf-musleh opened the pull request:
## Description

This PR serializes applied tags on various blocks and includes them in the OLX when they are copied. The format of the serialized tags are based on https://github.com/openedx/modular-learning/issues/174

## Supporting information

Related Ticket:
- https://github.com/openedx/modular-learning/issues/179
- Currently implemented without the use of Export ID, but should it should utilize it once the following PRs are merged:
    - https://github.com/openedx/openedx-learning/pull/145
    - https://github.com/openedx/edx-platform/pull/34143
- For the tags to serialize and appear for OpenAssessmentBlocks, we need:
    - https://github.com/openedx/edx-ora2/pull/2171

## Testing instructions

1. Start you localdev stack on this branch
1. Make sure you have Taxonomy/Tags data setup locally: https://github.com/open-craft/taxonomy-sample-data/
1. Once you have a course with a bunch of units and blocks with tags applied, trying copying units and various types of blocks
1. For each copy, double check that [staged content](http://localhost:18010/admin/content_staging/stagedcontent/) contains the tags data and there are no issues with pasting
1. Also confirm that when copying a unit/block that contains no tag, the `tags-v1` is not included in the OLX
1. Finally, make sure the tests cover the various cases.
1. For the OpenAssessmentBlock to work, you should setup https://github.com/openedx/edx-ora2/pull/2171 locally


---
Private-ref: [FAL-3605](https://tasks.opencraft.com/browse/FAL-3605)

yusuf-musleh committed:
feat: Serialize tag data in OLX for blocks

openedx-webhooks commented:
Thanks for the pull request, @yusuf-musleh! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

ChrisChV commented:
@yusuf-musleh I tested with an `openassessment` block and the tags are not saved in the OLX ([on the Sample Taxonomy Course > Unit 2.1.1](http://localhost:18010/container/block-v1:SampleTaxonomyOrg1+STC1+2023_1+type@vertical+block@2b9f9f4df8b34888830d9b037e24a61b))

bradenmacdonald commented on xmodule/xml_block.py:
```diff
@@ -498,6 +498,11 @@ def add_xml_to_node(self, node):
+        # Serialize and add tag data if any
+        tag_data = self.serialize_tag_data()
+        if tag_data:
+            node.set('tags-v1', tag_data)
```
```suggestion
        if isinstance(self, TaggedBlockMixin):
            tag_data = self.serialize_tag_data()
            if tag_data:
                node.set('tags-v1', tag_data)
```

I suggest you do this, then you don't have to update the LMS or all those test cases.

yusuf-musleh replied:
Update to this, it works well, thanks!

bradenmacdonald commented on setup.cfg:
```diff
@@ -168,6 +168,7 @@ ignore_imports =
     #  -> openedx.features.enterprise_support.utils
     openedx.features.enterprise_support.utils -> lms.djangoapps.branding.api
     cms.djangoapps.contentstore.rest_api.v1.views.settings -> lms.djangoapps.certificates.api
+    lms.envs.common -> cms.lib.xblock.tagging.tagged_block_mixin
```
See my other comment - I'd prefer not to add an exception here.

yusuf-musleh replied:
Removed 👍

bradenmacdonald commented on cms/lib/xblock/tagging/tagged_block_mixin.py:
```diff
@@ -0,0 +1,52 @@
+        Example serialized tags:
+            lightcast-skills:Typing,Microsoft Office;open-canada-skills:MS Office,%3Csome%3A%3B%2Cskill%2F%7C%3D%3E
+        """
+        from openedx.core.djangoapps.content_tagging.api import get_content_tags
```
Why is this import inside the function? Is there a circular import issue?

yusuf-musleh replied:
Since we are importing the `TaggedBlockMixin` in the cms settings and using it in the `XBLOCK_MIXINS`, the `content_tagging` app is not loaded yet, so having this defined outside causes a `django.core.exceptions.AppRegistryNotReady: Apps aren't loaded yet.` error. 

I added a comment mentioning this.

yusuf-musleh commented:
@ChrisChV I fixed the issued for OpenAssessmentBlocks, turns out we needed to make changes to the `edx-ora2` repo, not sure if that is the best approach or not, but it works. There's just some test coverage that is missing. To test it out, you'd need to setup https://github.com/openedx/edx-ora2/pull/2171 locally for it to work.

I've added a test case for it as well.

ChrisChV reviewed (approved):
@yusuf-musleh Thanks for this work! :+1:
- [X] I tested this: I followed the testing instructions.
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] Includes tests for bugfixes and/or features added.
- [X] Includes documentation

bradenmacdonald commented on requirements/edx/testing.txt:
```diff
@@ -992,7 +992,7 @@ optimizely-sdk==4.1.1
     #   -c requirements/edx/../constraints.txt
     #   -r requirements/edx/base.txt
-ora2==6.0.31
+ora2 @ git+https://github.com/open-craft/edx-ora2@yusuf-musleh/serialize-tag-data-in-openassessment-olx
```
Can you remove this change from the diff for now? We can merge them independently.

bradenmacdonald replied:
Oh I see, this has tests that depend on that PR. OK, we'll have to wait. Please ping me when the other PR has merged, and then I can merge this one.

yusuf-musleh replied:
Yup, will do.

yusuf-musleh commented:
@bradenmacdonald The ora2 PR has been merged, and the requirements in this PR are updated. It should be should be good to go.

edit: Accidentally clicked on close PR, my bad.

openedx-webhooks commented:
@yusuf-musleh 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34145&entry.1484655318=https://github.com/yusuf-musleh&entry.752974735=2024-01-30+16:08&entry.1917517419=2024-02-14+18:30&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

feanil commented on openedx/core/lib/xblock_serializer/test_api.py:
```diff
@@ -287,3 +415,273 @@ def test_jsinput_extra_files(self):
+            """
+        )
+
+    def test_tagged_openassessment_block(self):
```
@yusuf-musleh @bradenmacdonald @ChrisChV 

This test, relies on the openassessment block which comes from ora2 and having this test here caused an interesting issue.  When new attributes got added to the block in the edx-ora2 repo, it broke this test because the expected OLX in this repo did not have the new attributes.  This seems like bad coupling across repos and my instinct would be to move/remove this specific test.  But I'm not sure what the specific goals of this test and how important it is.  Can we remove this from the edx-platform repo?

yusuf-musleh replied:
@feanil The implementation/tests got updated since and now tags are no longer serialized and included in the actual OLX. So we could actually change the test case, namely changing [this assert](https://github.com/openedx/edx-platform/blob/master/openedx/core/lib/xblock_serializer/test_api.py#L704-L707) to just check that there is no tags data in the OLX, and we can get rid of `EXPECTED_OPENASSESSMENT_OLX` since we would no longer need it.

That way it wouldn't cause issues when attributes are added/removed from the ora2, and avoids removing the test case, does that sound like a reasonable approach?

feanil replied:
@yusuf-musleh that sounds great, please make that change and add a link to the PR here.

yusuf-musleh replied:
@feanil Sure thing, here it is https://github.com/openedx/edx-platform/pull/34782
